lib/stdlib/src/binary.erl

```diff
@@ -365,61 +365,124 @@ get_opts_replace(_,_) ->
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %% Hex encoding functions
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+-define(HEX(X), (hex(X)):16).
 -spec encode_hex(Bin) -> Bin2 when
       Bin :: binary(),
-      Bin2 :: binary().
-encode_hex(Bin) when is_binary(Bin) ->
-    encode_hex(Bin, <<>>);
+      Bin2 :: <<_:_*16>>.
+encode_hex(Data) when byte_size(Data) band 7 =:= 0 ->
```

Comment:
Shouldn't it be like that:

```suggestion
encode_hex(Data) when byte_size(Data) band 8 =:= 0 ->
```

Reply:
Don't think so. `Size band 7 =:= 0` means that the first three bits of the size are zeroes, that is, the size is not a multiple of any number from 1 to 7, i.e., the size is a multiple of 8. So we can digest in chunks of 8 bytes.
`Size band 8 =:= 0` means that the fourth bit of the size is a zero, that is, the size is not a multiple of 8, which breaks the purpose of this clause.

Reply:
Ok, I was tripped by the fact that this is only place where you use `band` instead of `rem`. Is there any performance penalty there? And if yes, then wouldn't it be better to use `byte_size(Data) band 7 =:= X` in other checks as well?

Reply:
That actually left me thinking how the beam handles these bitwise operations. In C there's reason to believe ands and ors to be theoretically faster than rems, unless the compiler is smart enough to see what's happening. Changing the and to a rem seems to worsen performance some 5% on my machine, weird. Wait, I'm going to experiment changing the logic of all clauses to use ands and ors 🤔

Reply:
> In C there's reason to believe ands and ors to be theoretically faster than rems, unless the compiler is smart enough to see what's happening. 

AFAIK any popular optimising compiler is "smart enough". If compilers can do things like changing between shifts/adds and multiply ([in both ways](https://youtu.be/bSkpMdDe4g4?t=1761)).

Reply:
Just played a bit with benchee, proper, and a good bunch of clause reorderings and bitwise operations, and they all seem to return pretty much exactly the same performance, variations are too minimal, probably a matter of noise (cpu getting hot or whatever). Compiler should handle that nice, so I'd just leave the most readable options here.

Reply:
I think that `rem` would be most readable, but `band` can be there as well as long as all heads will use the same approach.

Reply:
If you want to try to chase the last couple of nanoseconds, you may want to try and inline the hex function. Not sure if it will show up in the benchmarks, but it should make things faster and since the literal for the tuple is de-duplicated it won't increase the size by much.

```erlang
-compile({inline,[hex/1]}).
```

Reply:
I've just tried the inline and made sure the code-size doesn't grow. It does makes thing slightly faster, somewhere between 5-15% faster, so enough to say it's not benchmarking noises 🙂